Editor: Remove added tags from suggestion list

jeremeylduvall opened the pull request:
This is a shot at fixing #2205 by removing tags already added to the post from the suggestion list dropdown. I've been working at this for a bit, but just now got to a working example. 

**How it works**

I wanted to do all of the work in [`components/token-field/index.jsx`](https://github.com/Automattic/wp-calypso/blob/master/client/components/token-field/index.jsx) as I thought this would be a universal need. Items already added to the input field shouldn't appear in the suggestion list. I added a prop to the `SuggestionList` component that accepts the current value of `TokenField`. Then, I ran a check in `_renderSuggestions` when returning the list elements that checks if `this.props.tokenFieldValue.indexOf( suggestion ) === -1`. If so, it returns the item as an `li`.

The main issue I ran into was when all suggestions were added to the post. The `ul` would return empty, but it would still appear. I added another if block in the `render` statement that compares the intersection of `this.props.tokenFieldValue` (the current value of the input) and `this.props.suggestions`. If the length of that intersection is equal to the length of `this.props.suggestions`, the `ul` is removed. 

**To test**

1. Load up this branch and start a new post at http://calypso.localhost:3000/post/
2. Open the tag field. You should see all of your suggestions in the dropdown.
3. Start typing. Suggestion filtering should still work as expected.
4. Click on a tag to add it to your post. It should be removed from the suggestion list.
5. Add all tags in the suggestion dropdown to your post. The `ul` should be removed leaving no trace of a dropdown.
6. Remove tags from the post individually. They should slowly repopulate the tag suggestion dropdown.

**GIF**

![tags](https://cloud.githubusercontent.com/assets/7240478/14771425/d0c0eedc-0a47-11e6-89b9-aaff28789bdf.gif)

I still want to give this another run through, but cc @nylen for early feedback on the approach if you don't mind. My early attempts at doing this in [`post-editor/editor-tags/index.jsx`](https://github.com/Automattic/wp-calypso/blob/master/client/post-editor/editor-tags/index.jsx) proved futile.

jeremeylduvall commented:
The original issue I expected to run into actually isn't an issue due to how `intersection` is being used for the comparison of `this.props.tokenFieldValue` and `this.props.suggestions`. I'm going to flip this to "Needs Review." Any feedback welcome. If we need to do this elsewhere, happy to leave the exportable component alone and make this switch in post-editor/editor-tags/index.jsx.

nylen commented:
> If we need to do this elsewhere, happy to leave the exportable component alone

We should do this inside the `TokenField` component, but there is a better place to do it:  the `_getMatchingSuggestions` method in `index.jsx`.

Then it becomes a very simple fix, and you don't have to worry about an empty `ul` or comparing against the value again.

Let's also get a test case added for the new behavior.

jeremeylduvall commented:
Thanks @nylen. Much easier that way. Swapped out the previous solution for the following `if` block in `index.jsx`.

```
if ( match.length === 0 ) {
	each( suggestions, function( suggestion ) {
		if ( this.props.value.indexOf( suggestion ) === -1 ) {
			availableSuggestions.push( suggestion );
		}
	}.bind( this ) );

	suggestions = availableSuggestions;
}
```

That removes suggestions that already exist in the token field. Also updated the tests in `token-field/test/index.jsx`. Still works as originally pictured.

nylen commented on client/components/token-field/index.jsx:
```diff
@@ -386,7 +395,7 @@ var TokenField = React.createClass( {
 		}
 
-		return take( suggestions, this.props.maxSuggestions )
+		return take( suggestions, this.props.maxSuggestions );
```
Nice little fix here!

nylen commented on client/components/token-field/test/index.jsx:
```diff
@@ -138,6 +138,10 @@ describe( 'TokenField', function() {
 		} );
 
+		it( 'should remove already added tags from suggestions', function() {
+			expect( getSuggestionsHTML() ).not.to.contain( getTokensHTML );
```
This test does not work as intended - `getTokensHTML` is a function.

nylen commented:
This is looking better.  A couple of hints:

- Use lodash `difference` - this will remove the need for the `availableSuggestions` variable
- `match.length === 0` and `match.length > 0` are mutually exclusive, so you only need to use one of these conditions, and just use `else` instead of `else if`

As noted in above comments, the added test case will need some revision.

nylen commented on client/components/token-field/suggestions-list.jsx:
```diff
@@ -83,7 +83,6 @@ var SuggestionsList = React.createClass( {
 	},
 
 	_renderSuggestions: function() {
-
```
Not a big deal, but I wouldn't make this modification here - I think general practice is to only make cleanups like this in code that's very close to areas where you're already making other changes.

jeremeylduvall commented:
Updated the code to use `difference` instead of creating another variable. Also fixed the test so that it now works as expected. For the test to work, I manually set the tokens since our mock uses `foo` and `bar` as tokens, but [doesn't include them in the suggestion list](https://github.com/Automattic/wp-calypso/blob/master/client/components/token-field/test/lib/token-field-wrapper.jsx#L22) (i.e. the test was passing by default). To get a fair test, I set the tokens to `of` and `and` for this test in particular since those are included in the suggestion list. Appreciate the input @nylen.

nylen commented:
Looks good and works well.  Thanks for taking this on!
